# Bài tập 1

Nguyễn Đỗ Chí Thảo - 1712159

## Câu 1 (1 điểm)

a/ Nhà phát triển đã có các tham số tương ứng với từng loại tiền xu.
Ví dụ: Đồng 10 cent thì sẽ có kích thước bao nhiêu? Trọng lượng bao nhiêu?
Do đó, không cần phải xây dựng một mô hình, trải qua một quá trình học tập để tìm các tham số tương ứng với từng loại đồng xu.
Đã có sẵn những thông số cho sẵn. Ứng với kích thước x, trọng lượng y thì đó là đồng z cent => Not learning 

b/ Nhà phát triển chưa có thông tin chính xác về đồng xu. Họ chỉ có một lượng lớn các đồng xu đã được gán nhãn. Qua đó, họ sẽ rút trích các thông tin đặc trưng từ bộ dữ liệu. Trải qua quá trình học tập, sẽ thu được những thông số tương ứng với từng loại đồng xu. Từ đó, mới có thể xây dựng được mô hình phân lớp. => Supervised Learning

c/ Máy tính sẽ trải qua quá trình học tập với người chơi (hoặc chính nó) để từ đó đánh giá những bước đi, cách đi nào tối ưu, trạng thái k hiện tại nếu đi bước đi đó thì sẽ xảy ra trạng thái k+1 như thế nào. Từ đó đánh giá -1 hay +1 cho mỗi bước. Từ đó, tìm ra được bước đi tiếp theo cho mỗi trạng thái. Trải qua quá trình huấn luyện, máy sẽ tăng cường được khả năng dự đoán, đánh giá, máy càng học được nhiều nước đi   => Reinforcement Learning

D

## Câu 2 (1 điểm)

Điều kiện để áp dụng mô hình máy học vào một bài toán:
- Có tồn tại mô hình
- Không thể viết được công thức chính xác
- Có dữ liệu của bài toán

Xét bài tập của chúng ta

(i) Classifying numbers into primes and non-primes: </br>
>Trường hợp này, chúng ta có thể viết trực tiếp công thức để tìm ra số nguyên tố nên không cần thiết phải sử dụng dữ liệu để xây dựng mô hình phân loại: số nguyên tố / không phải là số nguyên tố 

(ii) Detecting potential fraud in credit card charges:
- Có tồn tại mô hình: một cách trực tiếp, chúng ta có thể nhận thấy rằng khi nào sẽ xảy ra gian lận trong tín dụng. Ví dụ: Ngày bình thường giao dịch của khách hàng chỉ giao động trong khoảng 5-10 triệu, tuy nhiên đến một hôm bỗng giao dịch 1-2 tỷ. Tuy không phải chắc chắn nhưng chúng ta sẽ cảm thấy rất có khả năng đây có thể là một hành vi gian lận trong tín dụng
- Không thể viết được công thức chính xác: Chúng ta không thể viết trực tiếp một hàm đa thức hay một hàm mũ ... xuống giấy cho bài toán này
- Có dữ liệu của bài toán: Chúng ta có thể sử dụng dữ liệu trong những lần giao dịch trước đây của khách hàng

(iii) Determining the time it would take a falling object to hit the ground: </br> 
>Trong trường hợp này, chúng ta có thể sử dụng các công thức vật lý để tìm ra thời gian vật rơi xuống chạm đất. Hay nói cách khác, đã tồn tại công thức để giải. Do đó, không cần phải sừ dụng mô hình máy học

(iv) Determining the optimal cycle for traffic lights in a busy intersection.
- Có tồn tại mô hình: một cách cảm quan chúng ta có cảm thấy rằng có thể khi nào thì thời gian đèn tín hiệu sẽ tối ưu. Chẳng hạn, vào giờ cao điểm thì mật độ giao thông khá cao nên có thể thời gian đèn tín hiệu đỏ dài hơn, đèn xanh ngắn lại; vào khoảng thời gian vắng người thì thời gian của gian đèn tín hiệu đỏ ngắn lại, đèn xanh tăng lên có thể tốt. Do chúng ra có nhận cả nhận được nên có tồn tại mô hình.
- Không thể viết được công thức chính xác: Chúng ta không thể viết trực tiếp một hàm đa thức hay một hàm mũ ... xuống giấy cho bài toán này
- Có dữ liệu của bài toán: Chúng ta có thể sử dụng dữ liệu trong lưu lượng giao thông để giải quyết. Ví dụ, với chu kỳ đèn tín hiệu như thế thì tình trạng giao thông như thế nào ( có kẹt xe, có tai nạn,... )


A

## Câu 3 (1 điểm)

Gọi A là biến cố chọn viên bi thứ nhất là bi đen 

Gọi B là biến cố chọn túi có 2 bi 

Nói cách khác, xác suất chọn được viên bi thứ 2 màu đen khi chọn được viên bi thứ nhất màu đen chính là xác xuất chọn được túi có 2 viên bi đen khi chọn được 1 viên bi đen. ( Bởi vì chỉ chọn được 2 viên bi màu đen khi chọn được túi này)

Áp dụng định lý Bayes:
    
P(A) = 1/2 * 1 + 1/2 * 1/2 = 3/4 

P(B) = 1/2 

P(A|B) = 1 

P(B|A) = P(A|B) * P(B) / P(A) = 2/3


D

## Câu 4 (1 điểm)

Xác suất chọn phải bi không phải màu đỏ: 1-0.55=0.45 

Chọn 10 viên bi: p = 0.45^10= 3.405 *10^(-4)


B

## Câu 5 (1 điểm)

Theo câu 4, ta có xác suất chọn phải bi không phải màu đỏ đối với 1 mẫu là p = 0.45

Xác suất chọn được 1 mẫu có bi màu đỏ: 1 - p 

Xác suất chọn cả 1000 mẫu có được bi màu đỏ : (1 - p)^1000 

Xác suất có ít nhất 1 mẫu trong 1000 mẫu có bi màu đỏ: 1 - (1 - p)^1000 = 0.289 (dùng phần bù)



C

## Câu 6 (1 điểm)

Gọi fi ( i thuộc {1, 2, 3, 4, 5, 6, 7, 8} ) là các giá trị đầu ra có (possible target functions). 

Ta có bảng như bên dưới


In [1]:
import pandas as pd
import numpy as np
df=pd.DataFrame([[0,0,0,0,1,1,1,1],
                 [0,0,1,1,0,0,1,1],
                 [0,1,0,1,0,1,0,1],
                 [0,1,1,0,1,'?','?','?'],
                 [0,1,1,0,1,'?','?','?'],
                 [0,1,1,0,1,0,0,0],
                 [0,1,1,0,1,0,0,1],
                 [0,1,1,0,1,0,1,0], 
                 [0,1,1,0,1,0,1,1], 
                 [0,1,1,0,1,1,0,0], 
                 [0,1,1,0,1,1,0,1], 
                 [0,1,1,0,1,1,1,0], 
                 [0,1,1,0,1,1,1,1]]).T 
  
df.columns = ["x1","x2","x3", "yn","gn","f1","f2","f3","f4","f5","f6","f7","f8"]
df


x1 x2 x3 yn gn f1 f2 f3 f4 f5 f6 f7 f8
0  0  0  0  0  0  0  0  0  0  0  0  0  0
1  0  0  1  1  1  1  1  1  1  1  1  1  1
2  0  1  0  1  1  1  1  1  1  1  1  1  1
3  0  1  1  0  0  0  0  0  0  0  0  0  0
4  1  0  0  1  1  1  1  1  1  1  1  1  1
5  1  0  1  ?  ?  0  0  0  0  1  1  1  1
6  1  1  0  ?  ?  0  0  1  1  0  0  1  1
7  1  1  1  ?  ?  0  1  0  1  0  1  0  1

[a] g returns 1 for all three points. 
Ta có bảng giá trị như sau:

In [2]:

df=pd.DataFrame([[0,0,0,0,1,1,1,1],
                 [0,0,1,1,0,0,1,1],
                 [0,1,0,1,0,1,0,1],
                 [0,1,1,0,1,'?','?','?'],
                 [0,1,1,0,1,'1','1','1'],
                 [0,1,1,0,1,0,0,0],
                 [0,1,1,0,1,0,0,1],
                 [0,1,1,0,1,0,1,0], 
                 [0,1,1,0,1,0,1,1], 
                 [0,1,1,0,1,1,0,0], 
                 [0,1,1,0,1,1,0,1], 
                 [0,1,1,0,1,1,1,0], 
                 [0,1,1,0,1,1,1,1]]).T 
  
df.columns = ["x1","x2","x3", "yn","gn","f1","f2","f3","f4","f5","f6","f7","f8"]
df

x1 x2 x3 yn gn f1 f2 f3 f4 f5 f6 f7 f8
0  0  0  0  0  0  0  0  0  0  0  0  0  0
1  0  0  1  1  1  1  1  1  1  1  1  1  1
2  0  1  0  1  1  1  1  1  1  1  1  1  1
3  0  1  1  0  0  0  0  0  0  0  0  0  0
4  1  0  0  1  1  1  1  1  1  1  1  1  1
5  1  0  1  ?  1  0  0  0  0  1  1  1  1
6  1  1  0  ?  1  0  0  1  1  0  0  1  1
7  1  1  1  ?  1  0  1  0  1  0  1  0  1

Score = 1 * 3 + 3 * 2 + 3 * 1 = 12

[b] g returns 0 for all three points.
Ta có bảng giá trị như sau:

In [3]:
df=pd.DataFrame([[0,0,0,0,1,1,1,1],
                 [0,0,1,1,0,0,1,1],
                 [0,1,0,1,0,1,0,1],
                 [0,1,1,0,1,'?','?','?'],
                 [0,1,1,0,1,'0','0','0'],
                 [0,1,1,0,1,0,0,0],
                 [0,1,1,0,1,0,0,1],
                 [0,1,1,0,1,0,1,0], 
                 [0,1,1,0,1,0,1,1], 
                 [0,1,1,0,1,1,0,0], 
                 [0,1,1,0,1,1,0,1], 
                 [0,1,1,0,1,1,1,0], 
                 [0,1,1,0,1,1,1,1]]).T 
  
df.columns = ["x1","x2","x3", "yn","gn","f1","f2","f3","f4","f5","f6","f7","f8"]
df

x1 x2 x3 yn gn f1 f2 f3 f4 f5 f6 f7 f8
0  0  0  0  0  0  0  0  0  0  0  0  0  0
1  0  0  1  1  1  1  1  1  1  1  1  1  1
2  0  1  0  1  1  1  1  1  1  1  1  1  1
3  0  1  1  0  0  0  0  0  0  0  0  0  0
4  1  0  0  1  1  1  1  1  1  1  1  1  1
5  1  0  1  ?  0  0  0  0  0  1  1  1  1
6  1  1  0  ?  0  0  0  1  1  0  0  1  1
7  1  1  1  ?  0  0  1  0  1  0  1  0  1

Score = 1 * 3 + 3 * 2 + 3 * 1 = 12

[c] g is the XOR function applied to x, i.e., if the number of 1s in x is odd, g
returns 1; if it is even, g returns 0. Ta có bảng giá trị như sau:

In [4]:
df=pd.DataFrame([[0,0,0,0,1,1,1,1],
                 [0,0,1,1,0,0,1,1],
                 [0,1,0,1,0,1,0,1],
                 [0,1,1,0,1,'?','?','?'],
                 [0,1,1,0,1,'0','0','1'],
                 [0,1,1,0,1,0,0,0],
                 [0,1,1,0,1,0,0,1],
                 [0,1,1,0,1,0,1,0], 
                 [0,1,1,0,1,0,1,1], 
                 [0,1,1,0,1,1,0,0], 
                 [0,1,1,0,1,1,0,1], 
                 [0,1,1,0,1,1,1,0], 
                 [0,1,1,0,1,1,1,1]]).T 
  
df.columns = ["x1","x2","x3", "yn","gn","f1","f2","f3","f4","f5","f6","f7","f8"]
df

x1 x2 x3 yn gn f1 f2 f3 f4 f5 f6 f7 f8
0  0  0  0  0  0  0  0  0  0  0  0  0  0
1  0  0  1  1  1  1  1  1  1  1  1  1  1
2  0  1  0  1  1  1  1  1  1  1  1  1  1
3  0  1  1  0  0  0  0  0  0  0  0  0  0
4  1  0  0  1  1  1  1  1  1  1  1  1  1
5  1  0  1  ?  0  0  0  0  0  1  1  1  1
6  1  1  0  ?  0  0  0  1  1  0  0  1  1
7  1  1  1  ?  1  0  1  0  1  0  1  0  1

Score = 1 * 3 + 3 * 2 + 3 * 1 = 12

[d] g returns the opposite of the XOR function: if the number of 1s is odd, it
returns 0, otherwise returns 1. Ta có bảng giá trị như sau:

In [5]:
print('Cau d\n')
df=pd.DataFrame([[0,0,0,0,1,1,1,1],
                 [0,0,1,1,0,0,1,1],
                 [0,1,0,1,0,1,0,1],
                 [0,1,1,0,1,'?','?','?'],
                 [0,1,1,0,1,'1','1','0'],
                 [0,1,1,0,1,0,0,0],
                 [0,1,1,0,1,0,0,1],
                 [0,1,1,0,1,0,1,0], 
                 [0,1,1,0,1,0,1,1], 
                 [0,1,1,0,1,1,0,0], 
                 [0,1,1,0,1,1,0,1], 
                 [0,1,1,0,1,1,1,0], 
                 [0,1,1,0,1,1,1,1]]).T 
  
df.columns = ["x1","x2","x3", "yn","gn","f1","f2","f3","f4","f5","f6","f7","f8"]
df

Cau d



x1 x2 x3 yn gn f1 f2 f3 f4 f5 f6 f7 f8
0  0  0  0  0  0  0  0  0  0  0  0  0  0
1  0  0  1  1  1  1  1  1  1  1  1  1  1
2  0  1  0  1  1  1  1  1  1  1  1  1  1
3  0  1  1  0  0  0  0  0  0  0  0  0  0
4  1  0  0  1  1  1  1  1  1  1  1  1  1
5  1  0  1  ?  1  0  0  0  0  1  1  1  1
6  1  1  0  ?  1  0  0  1  1  0  0  1  1
7  1  1  1  ?  0  0  1  0  1  0  1  0  1

Score = 1 * 3 + 3 * 2 + 3 * 1 = 12

E

## Câu 7 (1 điểm)

In [6]:
import numpy as np
def generate_target_w():
    """
    Generates target_w (the vector of parameters of f) 
    from two random, uniformly distributed points in [-1, 1] x [-1, 1].
    
    Returns
    -------
    target_w : numpy array, shape (3, 1) 
        The vector of parameters of f.
    """
    # Generate two points from a uniform distribution over [-1, 1]x[-1, 1]
    p1 = np.random.uniform(-1, 1, 2)
    p2 = np.random.uniform(-1, 1, 2)

    # Compute the target W from these two points
    target_w = np.array([p1[1]*p2[0] - p1[0]*p2[1], p2[1] - p1[1], p1[0] - p2[0]]).reshape((-1, 1))
    
    return target_w
def generate_data(N, target_w):
    """
    Generates a data set by generating random inputs and then using target_w to generate the 
    corresponding outputs.
    
    Parameters
    ----------
    N : int
        The number of examples.
    target_w : numpy array, shape (3, 1) 
        The vector of parameters of f.
    
    Returns
    -------
    X : numpy array, shape (N, 3)
        The matrix of input vectors (each row corresponds to an input vector); the first column of 
        this matrix is all ones.
    Y : numpy array, shape (N, 1)
        The vector of outputs.        
    """
    bad_data = True # `bad_data = True` means: data contain points on the target line 
                    # (this rarely happens, but just to be careful)
                    # -> y's of these points = 0 (with np.sign); 
                    #    we don't want this (y's of data must be -1 or 1)
                    # -> re-generate data until `bad_data = False`
    
    while bad_data == True:
        X = np.random.uniform(-1, 1, (N, 2))
        X = np.hstack((np.ones((N, 1)), X)) # Add 'ones' column
        Y = np.sign(np.dot(X, target_w))
        if (0 not in Y): # Good data
            bad_data = False
    
    return X, Y
def run_PLA(X, Y):
    """
    Runs PLA.
    
    Parameters
    ----------
    X : numpy array, shape (N, 3)
        The matrix of input vectors (each row corresponds to an input vector); the first column of 
        this matrix is all ones.
    Y : numpy array, shape (N, 1)
        The vector of outputs.
    
    Returns
    -------
    w : numpy array, shape (3, 1) 
        The vector of parameters of g.
    num_iterations : int
        The number of iterations PLA takes to converge.
    """
    w = np.zeros((X.shape[1], 1)) # Init w
    iteration = 0
    stop=False
    # TODO
    while True:
      # Duyet qua toan bo data
      for i in range(len(X)):
        # Tinh h(x) cho x[i]
        delta=np.sign(np.dot(X[i], w))
        #So sanh y[i], neu khac cap nhat w
        if delta[0]!=Y[i][0]:
          #Cap nhat w
          w=w+X[i].reshape((-1, 1)).dot(Y[i]).reshape((-1, 1))
          #Tang iteration
          iteration+=1
          #Tinh sai so 
          delta=Y-np.sign(np.dot(X, w))
          # Thoat vong lap neu sai so = 0
          if np.linalg.norm(delta)==0:
            stop=True
            break
      # Thoat vong lap
      if stop:
        break
    return w, iteration
def main(N): # You don't have to name this function "main", you can use other names
    """
    Parameters
    ----------
    N : int
        The number of training examples.
    """
    num_runs = 1000
    avg_num_iterations = 0.0 # The average number of iterations PLA takes to converge
    avg_test_err = 0.0 # The average test error of g - the final hypothesis picked by PLA
    
    for r in range(num_runs):
        # Generate target_w
        target_w = generate_target_w()
        
        # Generate training set
        X, Y = generate_data(N, target_w)
        
        # Run PLA to pick g
        w, num_iterations = run_PLA(X, Y)
        
        # Generate test set
        X_test, Y_test = generate_data(10000, target_w)
        
        # Test g
        test_err = np.mean(np.sign(np.dot(X_test, w)) != Y_test)
        
        # Update average values
        avg_num_iterations += (num_iterations * 1.0 / num_runs)
        avg_test_err += (test_err * 1.0 / num_runs)
    
    # Print results
    print('avg_num_iterations = %f' % (avg_num_iterations))
    print('avg_test_err = %f' % (avg_test_err))


In [7]:
# Run to find out answers for question 7 & 8! 
main(N=10) # We can use `main(10)`, but `main(N=10)` is clearer

avg_num_iterations = 10.063000
avg_test_err = 0.111251


In [8]:
main(N=100)

avg_num_iterations = 127.208000
avg_test_err = 0.013458


Theo kết quả ở trên, ta có số vòng lặp trung bình là 10.063000. Kết quả này gần với 15 nhất.

B

## Câu 8 (1 điểm)

Theo kết quả ở trên, ta có độ lỗi trung bình là 0.111251. Kết quả này gần 0.1 nhất

C

## Câu 9 (1 điểm)

Theo kết quả ở trên, ta có số vòng lặp trung bình là 127.208000. Kết quả này gần với 100 nhất.

B

## Câu 10 (1 điểm)

Theo kết quả ở trên, ta có độ lỗi trung bình là 0.013458. Kết quả này gần với 0.01 nhất.

B